### DISCUSSION TITLE
CLI command for creating Collection content

### DISCUSSION THREAD

**Discussion by mattstein** [withastro/roadmap#553] [Proposal]
### Body

# Summary

A CLI command (like `astro create`) that could create a new file in a Content Collection.

# Background & Motivation

Astro stands out to me for the smoothness of its development experience that thoughtfully irons out little frictions. `npx astro add [integration]` is a perfect example of this: it doesn’t just make it easy to tack on an integration, it shows me exactly what it’s doing. Saves me time and teaches me, which is a lovely thing.

Content Collections are great too, which got me thinking they could help not just with _validating_ Markdown that’s been created but helping create it in the first place!

Each time I create a new post within a Content Collection, I go through a few steps I wouldn’t have to with a CMS:

1. Create a new slug-friendly `.md` file. 
2. Copy and paste frontmatter from a neighboring `.md` file into this new one. (Because my memory is awful.)
3. Write the human-friendly title, then wipe clean whatever frontmatter I pasted and populate what I can while I’m thinking through the post.

This is not the end of the world, of course, but it’s a little set of mechanical barriers that keep me from jumping into writing and focusing on ideas. This effect is magnified for any other author or teammate that may not be as familiar with the site’s structure and content.

Since Content Collections know what shape a post should have, why not add a command that smooths out this process? It could take a collection name and title, and create a slug-friendly `.md` file in the right folder, pre-populated with relevant frontmatter and that human-friendly title.

# Goal

Take advantage of Content Collections and make it easier to start writing new content.

# Example

Let’s say we have this schema for a blog post collection:

```js
const blogPost = defineCollection({
  schema: z.object({
    title: z.string(),
    summary: z.string().optional(),
    pubDate: z.date().default(new Date()),
    draft: z.boolean().default(true),
  })
});
```

_I’m probably cheating thinking `new Date()` would be sufficient, but pretend with me for a moment._

## Minimal Example w/Single Argument

The only required argument should be a reference to the content collection if the project has more than one.

```
npx astro create blog
```

This would create a new “Untitled” post for the blog collection, pre-populating frontmatter based on the content schema:

```markdown
<!-- src/content/blog/untitled.md -->
---
title: Untitled
summary: 
pubDate: 2023-04-12
draft: true
---
```

## Example with Post Title

This would do the same thing, but starting with a post title. (Populating the first field in the content schema.)

```
npx astro create blog "My Wondrous Post"
```

```markdown
<!-- src/content/blog/my-wondrous-post.md -->
---
title: My Wondrous Post
summary: 
pubDate: 2023-04-12
draft: true
---
```

## Example with Post Title + Markdoc

And since Astro supports more than Markdown for Content Collections, there might as well be an option to specify the file format:

```
npx astro create blog "My Wondrous Post" mdoc
```

```markdown
<!-- src/content/blog/my-wondrous-post.mdoc -->
---
title: My Wondrous Post
summary: 
pubDate: 2023-04-12
draft: true
---
```

**Comment by mattstein** [comment]
FWIW I’m down to take a stab at implementation if this seems like a reasonable idea.

**Reply by yuxxeun** [reply]
Actually i have `shell script` for doing this kind of job, its woriking well with my nextjs example blog, but its still error when i use with `astro based-blog`, if u interested with it, maybe we can solve together

**Comment by Princesseuh** [comment]
I like this idea 😄

One technical issue I think you might hit is that I don't see how you could generate the actual file from the Zod schema. But I'm sure that's something that we could figure out.

**Comment by ematipico** [comment]
I also like this idea! This command would save a lot of time!

**Comment by kresnasatya** [comment]
I like this idea! It saves my time to create md file manually.

**Comment by coding-in-public** [comment]
Would love this idea! 

You could do something like…`npx astro new blog fileName.md` to generate a new fileName.md file in the blog collection with predefined schema.

Hugo SSG does this with the `new` CLI command and the archetype file (essentially their schema for content types). 
- https://gohugo.io/commands/hugo_new_content/
- https://gohugo.io/content-management/archetypes/

**Comment by maxchang3** [comment]
I like this idea.

During my migration from Hexo to Astro, I considered something similar. Hexo’s [hexo new](https://hexo.io/docs/commands#new) command is quite handy.

I explored a few ways to reuse `src/content/config.ts`, but it appears that this isn't technically feasible for now.

A universal CLI using Zod schema as a config could be a possible solution for now, or perhaps there's a possibility for a custom CLI command API.

**Comment by eduardocereto** [comment]
Wouldn't it make more sense to create content on astro db?

**Comment by arthur-fontaine** [comment]
I would prefer a command to create _anything_.

For example, in Symfony, we can do `symfony console make:entity`, in Adonis, we can do `node ace make:controller users`, in Laravel we can do `php artisan make:factory PostFactory`.

I think it's this CLI that makes them great frameworks. It's super easy and fast to develop complex applications with these frameworks.

In Astro, I'd like to write `npx astro make:action` to create a new action, `npx astro make:page` to create a new page, and, `npx astro make:collection` and `npx astro make:content blog` for the case of this discussion.

This would make creating applications in Astro even smoother.

**Comment by maxchang3** [comment]
For anyone interested in this feature, I've built a CLI called [newmd](https://github.com/maxchang3/newmd) that generates Markdown files with frontmatter based on a Zod schema, using [zod-empty](https://github.com/toiroakr/zod-empty) to fill in default values. 

While it doesn’t yet integrate seamlessly with Astro (which has been quite challenging 😅), it simplifies content creation by automating frontmatter setup.

**Comment by delucis** [comment]
Linking a similar proposal that was posted after this for reference: #1078

**Comment by iabrmv** [comment]
Just wanted to say I love this thread — I ran into the same friction when managing my own blog, and ended up building a CLI tool called [`blogue`](https://github.com/iabrmv/blogue).

It reads your content/config.ts, uses your Zod schemas, and helps you generate posts with valid frontmatter + auto-PRs using GitHub CLI.
```
npm install -g @blogue/cli
blogue new       # Create post
blogue publish   # PR + auto-merge
```

It supports:
✅ Multiple collections
✅ Accurate frontmatter based on Zod
✅ GitHub workflow automation

I made it for my own workflow, but happy to collaborate if others find it useful too.
